assets/js/wc-gateway-ppec-generate-cart.js

```diff
@@ -30,52 +30,67 @@
 			$( '#woo_pp_ec_button_product > *' ).css( 'pointer-events', 'none' );
 		} );
 
+	var variation_valid = true;
+	var fields_valid = true;
+	var update_button = function() {
+		$( '#woo_pp_ec_button_product' ).trigger( ( variation_valid && fields_valid ) ? 'enable' : 'disable' );
+	};
+
+	var validate_form = function() {
+		silent_validation = true;
+		fields_valid = true;
+		form.find( 'select:enabled, input:enabled, textarea:enabled' ).each( function() {
+			fields_valid = fields_valid && this.checkValidity();
+		} );
+		silent_validation = false;
+		update_button();
+	};
+
 	// It's a variations form, button availability should depend on its events
 	if ( $( '.variations_form' ).length ) {
-		$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
+		variation_valid = false;
 
 		$( '.variations_form' )
 		.on( 'show_variation', function( event, form, purchasable ) {
-			$( '#woo_pp_ec_button_product' ).trigger( purchasable ? 'enable' : 'disable' );
+			variation_valid = purchasable;
+			update_button();
 		} )
 		.on( 'hide_variation', function() {
-			$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
+			variation_valid = false;
+			update_button();
 		} );
 	}
 
-	var get_attributes = function() {
-		var select = $( '.variations_form' ).find( '.variations select' ),
-			data   = {},
-			count  = 0,
-			chosen = 0;
-
-		select.each( function() {
-			var attribute_name = $( this ).data( 'attribute_name' ) || $( this ).attr( 'name' );
-			var value	  = $( this ).val() || '';
-
-			if ( value.length > 0 ) {
-				chosen++;
-			}
-
-			count++;
-			data[ attribute_name ] = value;
-		} );
-
-		return {
-			'count'      : count,
-			'chosenCount': chosen,
-			'data'       : data
-		};
-	};
+	// Disable the button if there are invalid fields in the product page (like required fields from Product Addons)
+	var silent_validation;
+	var form = $( 'form.cart' );
+	form.get( 0 ).addEventListener( 'invalid', function( e ) {
+		if ( silent_validation ) {
+			e.preventDefault();
+		}
+	}, true );
+	form.on( 'change', 'select, input, textarea', validate_form );
+	validate_form();
 
 	var generate_cart = function( callback ) {
 		var data = {
-			'nonce':       wc_ppec_generate_cart_context.generate_cart_nonce,
-			'qty':         $( '.quantity .qty' ).val(),
-			'attributes':  $( '.variations_form' ).length ? get_attributes().data : [],
-			'add-to-cart': $( '[name=add-to-cart]' ).val(),
+			'nonce': wc_ppec_generate_cart_context.generate_cart_nonce,
 		};
 
+		var field_pairs = form.serializeArray();
+		for ( var i = 0; i < field_pairs.length; i++ ) {
+			// Prevent the default WooCommerce PHP form handler from recognizing this as an "add to cart" call
+			if ( 'add-to-cart' === field_pairs[ i ].name ) {
+				field_pairs[ i ].name = 'ppec-add-to-cart';
+			}
+			data[ field_pairs[ i ].name ] = field_pairs[ i ].value;
+		}
+
+		if ( ! data[ 'ppec-add-to-cart' ] ) {
```

Comment:
Why are we checking for the existence of `data[ 'ppec-add-to-cart' ]` first?

Reply:
Hum... I suppose that's redundant, there's no harm in always overwriting `data[ 'ppec-add-to-cart' ]` even if the product is a variable product. Removed in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/507/commits/dc8c28341409226ea4a069b794eaa159ed10c26c